### DISCUSSION TITLE
saved_path not defined in Jupyter notebook on Windows

### DISCUSSION THREAD

**Discussion by yashika51** [bentoml/BentoML#931] [General] [answered]
Hi,
As far as it seems it is a problem due to slashes (`/`) on windows. I am starting a discussion here so that other windows users can also refer it in the future.

While loading the saved model in jupyter notebooks(vs code) an error stating `saved_path` not defined occurs.
 
![image](https://user-images.githubusercontent.com/34909206/88244150-87084d00-ccb0-11ea-91bc-b69992e3e8c5.png)

As a workaround when I print the complete path and pass that to load the model another error `config file yml not found occurs`. 
![image](https://user-images.githubusercontent.com/34909206/88244110-532d2780-ccb0-11ea-867f-c1af20440753.png)

I checked that the file exists but the function is not able to track that. 
However, when I run the same code as a python script instead of notebook, I get the desired result. Something similar happens with another person from my pod.

What workaround can be there for using bentoml's load function+jupyter notebooks on Windows?

**Comment by parano** [comment]
Hi @yashika51 - could you share your notebook?  there are many windows users using bentoML and I haven't seen this issue came up before.  Based on the screenshot, it looks like you used `%%writefile` to create the main.py file instead of running that piece of code.

**Comment by yashika51** [comment, marked as answer]
Thanks for the reply @parano. I referred to examples from `bentoml/gallery` which were using `%writefile` method. 
Here is the notebook:
https://nbviewer.jupyter.org/gist/yashika51/7635e3120a92869658611d410f33e83e

**Reply by parano** [reply]
I see - so the reason you are seeing this error, is because when you do `!python main.py`, the code is being executed in a new python session instead of the current session. That's why you can not access the `saved_path` variable after running `!python main.py`.

The first cell should use `%%writefile` because BentoML requires putting the main BentoService class definition to its own python file. Defining BentoService class in the `main` module(running in a python interactive shell or notebook) makes it hard to do code packaging:
```python
%%writefile iris_classifier.py
#writing script for BentoML Service Class
from bentoml import env,artifacts,api, BentoService
from bentoml.artifact import SklearnModelArtifact
from bentoml.adapters import DataframeInput

@env(auto_pip_dependencies=True,
     conda_dependencies=['bentoml'])
@artifacts([SklearnModelArtifact("model")])
class IrisClassifier(BentoService):
    @api(input=DataframeInput())
  
    def predict(self,df):
        return self.artifact.model.predict(df)
```

However you don't need the `%%writefile` for the second cell, you can simply change the second cell to:
```python

    from sklearn.svm import SVC
    from sklearn.datasets import load_iris
    from iris_classifier import IrisClassifier 

    #load data
    iris=load_iris()
    X,y=iris.data, iris.target

    #train model
    clf=SVC()
    clf.fit(X,y)

    #create an iris classifier service instance 
    service=IrisClassifier()

    #pack the newly trained service classifier
    service.pack("model",clf)

    # Save the prediction service to disk for model serving
    saved_path=service.save()
```

That will execute all code in current notebook's python session and give you access to the `saved_path` variable in following cells.

**Reply by yashika51** [reply]
Oh, Understood. Thanks for explaining that. It works fine now 😄. Should we do this small fix in the code examples? That is removing `%%writefile` from the second cell?

https://github.com/bentoml/gallery/blob/master/scikit-learn/iris-classifier/iris-classifier.ipynb
![image](https://user-images.githubusercontent.com/34909206/88469352-60dfe880-cf0e-11ea-871a-d2bfba2e4b97.png)

**Reply by parano** [reply]
Yea if you could help with a quick fix that'd be great. That example uses the `!bentoml serve IrisClassifier:latest` CLI command to load the service instead of using the `saved_path` variable so it runs without problem. But I agree it is unnecessary to write that code in a `main.py` file, it can be very confusing to new users.

**Reply by yashika51** [reply]
I agree. I'll make a PR for this small fix. :)
